source/isaaclab/isaaclab/sensors/tacsl_sensor/gelsight_render.py

```diff
@@ -0,0 +1,234 @@
+# Copyright (c) 2022-2025, The Isaac Lab Project Developers (https://github.com/isaac-sim/IsaacLab/blob/main/CONTRIBUTORS.md).
+# All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+import numpy as np
+import os
+import scipy
+import torch
+
+import cv2
+
+from isaaclab.utils.assets import ISAACLAB_NUCLEUS_DIR, retrieve_file_path
+
+conf_r15 = {
+    "data_dir": "gelsight_r15_data",
+    "background_path": "bg.jpg",
+    "calib_path": "polycalib.npz",
+    "real_bg": "real_bg.npy",
+    "h": 320,
+    "w": 240,
+    "numBins": 120,
+    "pixmm": 0.0877,
+}
+conf_gs_mini = {
+    "data_dir": "gs_mini_data",
+    "background_path": "bg.jpg",
+    "calib_path": "polycalib.npz",
+    "real_bg": "real_bg.npy",
+    "h": 240,
+    "w": 320,
+    "numBins": 120,
+    "pixmm": 0.065,
+}
+conf_options = {
+    "gelsight_r15": conf_r15,
+    "gs_mini": conf_gs_mini,
+}
+
+
+def generate_normals_tensor(img_tensor):
+    """
+    Generate the gradient magnitude and direction of the height map using tensors.
+
+    Parameters:
+    img_tensor (torch.Tensor): Input height map tensor.
+
+    Returns:
+    tuple: Gradient magnitude tensor, gradient direction tensor, and None.
+    """
+    img_grad_tensor = torch.gradient(img_tensor, dim=(1, 2))
+    dzdx, dzdy = img_grad_tensor
+
+    grad_mag_orig_tensor = torch.sqrt(dzdx**2 + dzdy**2)
+    grad_mag_tensor = torch.arctan(grad_mag_orig_tensor)  # seems that arctan is used as a squashing function
+    grad_dir_tensor = torch.arctan2(dzdx, dzdy)
+    grad_dir_tensor[grad_mag_orig_tensor == 0] = 0
+
+    # handle edges
+    grad_mag_tensor = torch.nn.functional.pad(grad_mag_tensor[:, 1:-1, 1:-1], pad=(1, 1, 1, 1))
+    grad_dir_tensor = torch.nn.functional.pad(grad_dir_tensor[:, 1:-1, 1:-1], pad=(1, 1, 1, 1))
+
+    return grad_mag_tensor, grad_dir_tensor, None
+
+
+def get_filtering_kernel(kernel_sz=5):
+    """
+    Create a Gaussian filtering kernel.
+    # For kernel derivation, see https://cecas.clemson.edu/~stb/ece847/internal/cvbook/ch03_filtering.pdf
+
+    Parameters:
+    kernel_sz (int): Size of the kernel.
+
+    Returns:
+    numpy.ndarray: Filtering kernel.
+    """
+    filter_1D = scipy.special.binom(kernel_sz - 1, np.arange(kernel_sz))
+    filter_1D /= filter_1D.sum()
+    filter_1D = filter_1D[..., None]
+
+    kernel = filter_1D @ filter_1D.T
+    return kernel
+
+
+def gaussian_filtering(img_tensor, kernel_tensor):
```

Comment:
Are users expected to call these functions directly? If not, I think it is better to move them inside the GelSightRender class and mark them as a private using an underscore.